components/clm/src/biogeochem/LSparseMatMod.F90

```diff
@@ -0,0 +1,420 @@
+    ! !ARGUMENTS:
+    class(lom_type), intent(in) :: this
+    integer , intent(in) :: nprimvars
+    integer , intent(in) :: nr
+    real(r8), intent(in) :: pscal(1:nprimvars)
+    type(sparseMat_type), intent(in) :: spm_d
+    real(r8), intent(out):: rscal(1:nr)
+    ! !LOCAL VARIABLES:
+    integer :: jj, ii, id
+
+    rscal(:)=1._r8
+    do ii=1, spm_d%szrow
+      do jj = 1, spm_d%ncol(ii)
+        id = spm_d%pB(ii)+jj-1
+        rscal(spm_d%icol(id)) = min(pscal(ii),rscal(spm_d%icol(id)))
+      enddo
+    enddo
+
+  end subroutine calc_reaction_rscal
+
+  !-------------------------------------------------------------------------------
+  subroutine apply_reaction_rscal(this, nr, rscal, reaction_rates)
+    !
+    ! !DESCRIPTION:
+    ! reduce reaction rates using input scalar
+    !
+    implicit none
+    ! !ARGUMENTS:
+    class(lom_type), intent(in) :: this
+    integer , intent(in)    :: nr
+    real(r8), intent(in)    :: rscal(1:nr)
+    real(r8), intent(inout) :: reaction_rates(1:nr)
+    ! !LOCAL VARIABLES:
+    integer :: j
+
+    do j = 1, nr
+       reaction_rates(j) = reaction_rates(j)*rscal(j)
+    enddo
+  end subroutine  apply_reaction_rscal
+
+  !-------------------------------------------------------------------------------
+  subroutine flux_correction(nvars, nreactions, spm_p, spm_d, dtime, ystates, rfluxes)
+  !
+  ! DESCRIPTION
+  ! correcting the fluxes to avoid negative state variables
+  use abortutils             , only : endrun
+  implicit none
+  integer, intent(in) :: nvars
+  integer, intent(in) :: nreactions
+  class(sparseMat_type), intent(in) :: spm_p
+  class(sparseMat_type), intent(in) :: spm_d
+  real(r8), intent(in) :: dtime
+  real(r8), intent(in) :: ystates(nvars)
+  real(r8), intent(inout):: rfluxes(nreactions)
+
+  integer :: it
+  real(r8) :: rscal(nreactions)
+  real(r8) :: pscal(nvars)
+  real(r8) :: d_dt(nvars)
+  real(r8) :: p_dt(nvars)
+  integer :: errinfo
+  type(lom_type) :: lom
+  logical :: lneg
+  integer, parameter :: itmax=40
+  it=0
+  rscal=0._r8
+  do
+    !obtain the destruction fluxes
+    d_dt(:)=0._r8
+    call spm_axpy(spm_d%szrow, spm_d%szcol, 1._r8, rfluxes, spm_d, d_dt, errinfo)
+
+    if(errinfo<0)then
+      call endrun(msg='ERROR:: in spm_axpy'//errMsg(__FILE__, __LINE__))
+    endif
+    !obtain the production fluxes
+    p_dt(:)=0._r8
+    call spm_axpy(spm_p%szrow, spm_d%szcol, 1._r8, rfluxes, spm_p, p_dt, errinfo)
+
+    if(errinfo<0)then
+      call endrun(msg='ERROR:: in spm_axpy '//errMsg(__FILE__, __LINE__))
+    endif
+    call lom%calc_state_pscal(spm_d%szrow, dtime, ystates, p_dt,  d_dt, &
+      pscal, lneg, errinfo)
+    if(errinfo<0)then
+      call endrun(msg='ERROR:: in calc_state_pscal '//errMsg(__FILE__, __LINE__))
+    endif
+    if(lneg .and. it<=itmax)then
+      call lom%calc_reaction_rscal(spm_d%szrow, spm_d%szcol,  pscal, &
+        spm_d,rscal)
+
+      call lom%apply_reaction_rscal(spm_d%szcol, rscal, rfluxes)
+    else
+      exit
```

Comment:
Should a warning be printed that the flux correction failed to converge after `itmax` iterations?

Reply:
One warning print out is added, although my experience suggests on exiting at maximum iteration, even the answer is still negative, its absolute variable is less than roundoff error and is zeroed off in precision control.